File: Js/expression-evaluate.js
function getOperatorSuperiority(operator) {
    if (operator === '+' || operator === '-' || operator === '**') return 1;
    else if (operator === '*' || operator === '/') return 2;
    return 0;
}

function operations(value_one, value_two, operator) {
    switch (operator) {
        case '+': return value_one + value_two;
        case '-': return value_one - value_two;
        case '*': return value_one * value_two;
        case '/': return value_one / value_two;
        case '**': return value_one ** value_two;
    }
}

function removeValuesAndOperator(operands, operators) {
    let value_one = operands.pop();
    let value_two = operands.pop();
    let operator = operators.pop();
    return operations(Number(value_two), Number(value_one), operator);
}

function tokenize(s) {
    return s.replace(/\(/g, " ( ").replace(/\)/g, " ) ").trim().split(/\s+/);
}

function expressionEvaluate(expression) {
    const operators = [], values = [];
    const tokens = tokenize(expression);

    for (let index = 0; index < tokens.length; index++) {
        if (Number(tokens[index])) values.push(tokens[index]);
        else if (tokens[index] === '(') operators.push(tokens[index]);
        else if (operators.length === 0) operators.push(tokens[index]);
        else if (tokens[index] === ')') {
            while (operators[operators.length - 1] !== '(') {
                values.push(removeValuesAndOperator(values, operators));
            }
            operators.pop();
        } else if (
            getOperatorSuperiority(operators[operators.length - 1]) <= getOperatorSuperiority(tokens[index])) {
            operators.push(tokens[index]);
        } else if (getOperatorSuperiority(operators[operators.length - 1]) >= getOperatorSuperiority(tokens[index])) {
            values.push(removeValuesAndOperator(values, operators));
            operators.push(tokens[index]);
        }
    }
    while (operators.length > 0) {
        values.push(removeValuesAndOperator(values, operators));
    }
    return values;
}

console.log('Result = 10', expressionEvaluate('5 + 5')); // 10
console.log('Result = 5', expressionEvaluate('10 - 5')); // 5
console.log('Result = 20', expressionEvaluate('2 * 10')); // 20
console.log('Result = 5', expressionEvaluate('10 / 2')); // 5
console.log('Result = 27', expressionEvaluate('3 ** 3')); // 27
console.log('Result = 18', expressionEvaluate('5 + 5 * 2 + 3')); // 18
console.log('Result = 15', expressionEvaluate('5 * 2 + (4 + 1)')); // 15
console.log('Result = 10', expressionEvaluate('10 / 2 + (5 * 1)')); // 10
console.log('Result = 14', expressionEvaluate('28 / 2')); // 14
console.log('Result = 7', expressionEvaluate('10 / 2 + 2')); // 7

File: Python/expression-evaluate.py
def operatorWithGreaterSuperiority(operator):
    if operator == '+' or operator == '-' or operator == '**':
        return 1
    elif operator == '*' or operator == '/': 
        return 2
    else:
        return 0

def operations(value_one, value_two, operator):
    if operator == '+':
        return value_one + value_two
    elif operator == '-':
        return value_one - value_two
    elif operator == '*': 
        return value_one * value_two
    elif operator == '/':
        return value_one / value_two
    elif operator == '**':
        return value_one ** value_two
    else:
        return 0

def removeValuesAndOperator(operands, operators):
    value_one = operands.pop()
    value_two = operands.pop()
    operator = operators.pop()
    return operations(value_two, value_one, operator)

def transformDigitInNumber(number):
    transform = 0
    transform = (transform * 10)  + int(number)
    return transform

def expressionEvaluate(expression):
    operators = []
    values = []
    expression = expression.split(' ')
    for index, item in enumerate(expression):
        if expression[index].isdigit():
            values.append(transformDigitInNumber(expression[index]))
        elif expression[index] == '(':
            operators.append(item)
        elif len(operators) == 0:
            operators.append(item)
        elif expression[index] == ')':
            while len(operators) > 0 and operators[-1] != '(':
                values.append(removeValuesAndOperator(values, operators))
            operators.pop()
        elif operatorWithGreaterSuperiority(operators[-1]) <= operatorWithGreaterSuperiority(expression[index]):
            operators.append(item)          
        elif operatorWithGreaterSuperiority(operators[-1]) >= operatorWithGreaterSuperiority(expression[index]): 
                values.append(removeValuesAndOperator(values, operators))
                operators.append(item)
    while len(operators) > 0:
            values.append(removeValuesAndOperator(values, operators))
    return values

if __name__ == "__main__":
    print(f'Result 25: {expressionEvaluate("5 * 5")}')
    print(f'Result 12: {expressionEvaluate("6 + 6")}')
    print(f'Result 3: {expressionEvaluate("6 - 3")}')
    print(f'Result 3: {expressionEvaluate("6 / 2")}')
    print(f'Result 17: {expressionEvaluate("3 * 3 + ( 4 + 4 )")}')
    print(f'Result 56: {expressionEvaluate("50 + 2 * 3")}')
    print(f'Result 100: {expressionEvaluate("10 ** 2")}')
    print(f'Result 70: {expressionEvaluate("10 + 6 * 10")}')



